crates/mun_codegen/src/code_gen.rs

```diff
@@ -34,87 +39,122 @@ impl From<LinkerError> for CodeGenerationError {
-    // cached value.
-    let module = db.module_ir(file_id);
-    assembly_module
-        .link_in_module(module.llvm_module.clone())
-        .map_err(|e| CodeGenerationError::ModuleLinkerError(e.to_string()))?;
-
-    // Generate the `get_info` method.
-    symbols::gen_reflection_ir(
-        db,
-        &target_machine,
-        &assembly_module,
-        &module.types,
-        &module.functions,
-        &module.dispatch_table,
-    );
-
-    // Optimize the assembly module
-    optimize_module(&assembly_module, db.optimization_lvl());
-
-    // Debug print the IR
-    //println!("{}", assembly_module.print_to_string().to_string());
-
-    // Generate object file
-    let obj_file = {
-        let obj = target_machine
-            .write_to_memory_buffer(&assembly_module, FileType::Object)
-            .map_err(|e| CodeGenerationError::CodeGenerationError(e.to_string()))?;
-        let mut obj_file = tempfile::NamedTempFile::new()
-            .map_err(CodeGenerationError::CouldNotCreateObjectFile)?;
-        obj_file
-            .write(obj.as_slice())
-            .map_err(CodeGenerationError::CouldNotCreateObjectFile)?;
-        obj_file
-    };
-
-    // Construct a linker for the target
-    let mut linker = linker::create_with_target(&target);
-    linker.add_object(obj_file.path())?;
-
-    // Link the object
-    linker.build_shared_object(&output_file_path)?;
-    linker.finalize()?;
-
-    Ok(())
+    _target: inkwell::targets::Target,
+    target_machine: inkwell::targets::TargetMachine,
+    assembly_module: Arc<inkwell::module::Module>,
+}
+
+impl<'a, D: IrDatabase> ModuleBuilder<'a, D> {
+    /// Construct module for the given `hir::FileId` at the specified output file location.
+    pub fn new(db: &'a mut D, file_id: FileId) -> Result<Self, failure::Error> {
+        let target = db.target();
+
+        // Construct a module for the assembly
+        let assembly_module = Arc::new(
+            db.context()
+                .create_module(db.file_relative_path(file_id).as_str()),
+        );
+
+        // Initialize the x86 target
+        Target::initialize_x86(&InitializationConfig::default());
+
+        // Retrieve the LLVM target using the specified target.
+        let llvm_target = Target::from_triple(&target.llvm_target)
+            .map_err(|e| CodeGenerationError::UnknownTargetTriple(e.to_string()))?;
+        assembly_module.set_target(&llvm_target);
+
+        // Construct target machine for machine code generation
+        let target_machine = llvm_target
+            .create_target_machine(
+                &target.llvm_target,
+                &target.options.cpu,
+                &target.options.features,
+                db.optimization_lvl(),
+                RelocMode::PIC,
+                CodeModel::Default,
+            )
+            .ok_or(CodeGenerationError::CouldNotCreateTargetMachine)?;
+
+        // Initialize the module and target data
+        db.set_module(assembly_module.clone());
```

Comment:
This is basically the equivalent of setting global state which is very unsafe if the code is called from multiple threads at the same time. We have to find a better solution.

It might be better to NOT use salsa in this case because I don't think we're winning a lot with it in this case. Especially since the data we store in it is not truley clonable.